lib/librte_hash/rte_cuckoo_hash.c

```diff
@@ -1565,6 +1774,146 @@ rte_hash_del_key_with_hash(const struct rte_hash *h,
 	return __rte_hash_del_key_with_hash(h, key, sig);
 }
 
+/* Search one bucket and remove the matched key and yield.
+ * Writer is expected to hold the lock while calling this
+ * function.
+ */
+static inline int32_t
+search_and_remove_and_yield(const struct rte_hash *h, const void *key,
+	struct rte_hash_bucket *bkt, uint16_t sig, int *pos,
+	rte_hash_yield_func yield_func, void *arg)
+{
+	struct rte_hash_key *k, *keys = h->key_store;
+	unsigned int i;
+	uint32_t key_idx;
+
+	/* Check if key is in bucket */
+	for (i = 0; i < RTE_HASH_BUCKET_ENTRIES; i++) {
+		key_idx = __atomic_load_n(&bkt->key_idx[i],
+					  __ATOMIC_ACQUIRE);
+		if (bkt->sig_current[i] == sig && key_idx != EMPTY_SLOT) {
+			k = (struct rte_hash_key *) ((char *)keys +
+					key_idx * h->key_entry_size);
+			yield_func(k, arg);
+			if (rte_hash_cmp_eq(key, k->key, h) == 0) {
+				bkt->sig_current[i] = NULL_SIGNATURE;
+				/* Free the key store index if
+				 * no_free_on_del is disabled.
+				 */
+				if (!h->no_free_on_del)
+					remove_entry(h, bkt, i);
+
+				__atomic_store_n(&bkt->key_idx[i],
+						 EMPTY_SLOT,
+						 __ATOMIC_RELEASE);
+
+				*pos = i;
+				/*
+				 * Return index where key is stored,
+				 * subtracting the first dummy index
+				 */
+				return key_idx - 1;
+			}
+		}
+	}
+	return -1;
+}
+
+static inline int32_t
+__rte_hash_del_key_and_yield_with_hash(const struct rte_hash *h,
+	const void *key, hash_sig_t sig, rte_hash_yield_func yield_func,
+	void *arg)
+{
+	uint32_t prim_bucket_idx, sec_bucket_idx;
+	struct rte_hash_bucket *prim_bkt, *sec_bkt, *prev_bkt, *last_bkt;
+	struct rte_hash_bucket *cur_bkt;
+	int pos;
+	int32_t ret, i;
+	uint16_t short_sig;
+
+	short_sig = get_short_sig(sig);
+	prim_bucket_idx = get_prim_bucket_index(h, sig);
+	sec_bucket_idx = get_alt_bucket_index(h, prim_bucket_idx, short_sig);
+	prim_bkt = &h->buckets[prim_bucket_idx];
+	yield_func(prim_bkt, arg);
+
+	__hash_rw_writer_lock(h);
+	/* look for key in primary bucket */
+	ret = search_and_remove(h, key, prim_bkt, short_sig, &pos);
+	if (ret != -1) {
+		__rte_hash_compact_ll(h, prim_bkt, pos);
+		last_bkt = prim_bkt->next;
+		prev_bkt = prim_bkt;
+		goto return_bkt;
+	}
+
+	/* Calculate secondary hash */
+	sec_bkt = &h->buckets[sec_bucket_idx];
+
+	FOR_EACH_BUCKET(cur_bkt, sec_bkt) {
+		yield_func(cur_bkt, arg);
+		ret = search_and_remove(h, key, cur_bkt, short_sig, &pos);
+		if (ret != -1) {
+			__rte_hash_compact_ll(h, cur_bkt, pos);
+			last_bkt = sec_bkt->next;
+			prev_bkt = sec_bkt;
+			goto return_bkt;
+		}
+	}
+
+	__hash_rw_writer_unlock(h);
+	return -ENOENT;
+
+/* Search last bucket to see if empty to be recycled */
+return_bkt:
+	if (!last_bkt) {
+		__hash_rw_writer_unlock(h);
+		return ret;
+	}
+	while (last_bkt->next) {
+		prev_bkt = last_bkt;
+		last_bkt = last_bkt->next;
+	}
+
+	yield_func(last_bkt, arg);
```

Comment:
This function should come before the `while` loop above AND as the last line of the `while` loop.